From 07cbb78aff934f4d718cc155aaecf6678ade76ce Mon Sep 17 00:00:00 2001
From: yuanhecai <yuanhecai@loongson.cn>
Date: Mon, 5 Feb 2024 17:06:16 +0800
Subject: [PATCH 02/15] BACKPORT: FROMLIST: LoongArch64: Add loongarch support

Lsx and Lasx optimizations are automatically enabled by detecting
the characteristics of the CPU.

Change-Id: I22de05163adb075856722d37f15ac8469ba5c6b0

Link: https://bitbucket.org/multicoreware/x265_git/pull-requests/31
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 source/CMakeLists.txt        | 45 ++++++++++++++++++++++++++++++++++--
 source/cmake/FindLASX.cmake  | 14 +++++++++++
 source/cmake/FindLSX.cmake   | 14 +++++++++++
 source/common/CMakeLists.txt |  7 ++++--
 source/common/cpu.cpp        | 31 +++++++++++++++++++++++--
 source/test/testbench.cpp    |  2 ++
 source/x265.h                |  4 ++++
 7 files changed, 111 insertions(+), 6 deletions(-)
 create mode 100644 source/cmake/FindLASX.cmake
 create mode 100644 source/cmake/FindLSX.cmake

diff --git a/source/CMakeLists.txt b/source/CMakeLists.txt
index ab5ddfeb7..785015ec1 100755
--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -51,6 +51,9 @@ list(FIND ARM_ALIASES "${SYSPROC}" ARMMATCH)
 list(FIND ARM64_ALIASES "${SYSPROC}" ARM64MATCH)
 set(POWER_ALIASES powerpc64 powerpc64le ppc64 ppc64le)
 list(FIND POWER_ALIASES "${SYSPROC}" POWERMATCH)
+set(LOONGARCH64_ALIASES loongarch64)
+list(FIND LOONGARCH64_ALIASES "${SYSPROC}" LOONGARCH64MATCH)
+
 if(X86MATCH GREATER "-1")
     set(X86 1)
     add_definitions(-DX265_ARCH_X86=1)
@@ -88,6 +91,10 @@ elseif(ARM64MATCH GREATER "-1")
     message(STATUS "Detected ARM64 target processor")
     set(ARM64 1)
     add_definitions(-DX265_ARCH_ARM64=1 -DHAVE_NEON)
+elseif(LOONGARCH64MATCH GREATER "-1")
+    set(LOONGARCH64 1)
+    add_definitions(-DLOONGARCH64=1 -DX265_ARCH_LOONGARCH64=1)
+    message(STATUS "Detected LOONGARCH64 target processor")
 else()
     message(STATUS "CMAKE_SYSTEM_PROCESSOR value `${CMAKE_SYSTEM_PROCESSOR}` is unknown")
     message(STATUS "Please add this value near ${CMAKE_CURRENT_LIST_FILE}:${CMAKE_CURRENT_LIST_LINE}")
@@ -286,6 +293,23 @@ if(GCC)
 	list(APPEND ARM_ARGS -DPIC)
 	endif()
     add_definitions(${ARM_ARGS})
+    if(LOONGARCH64)
+        set(LOONGARCH64_ARGS -fPIC)
+
+        find_package(LSX)
+        find_package(LASX)
+
+        if(SUPPORTS_LSX)
+            add_definitions(-DHAVE_LSX=1)
+        endif()
+
+        if(SUPPORTS_LASX)
+            add_definitions(-DHAVE_LASX=1)
+        endif()
+
+    endif()
+    add_definitions(${LOONGARCH64_ARGS})
+
     if(FPROFILE_GENERATE)
         if(INTEL_CXX)
             add_definitions(-prof-gen -prof-dir="${CMAKE_CURRENT_BINARY_DIR}")
@@ -410,7 +434,7 @@ if(EXTRA_LIB)
 endif(EXTRA_LIB)
 mark_as_advanced(EXTRA_LIB EXTRA_LINK_FLAGS)
 
-if(X64 OR ARM64 OR PPC64)
+if(X64 OR ARM64 OR PPC64 OR LOONGARCH64)
     # NOTE: We only officially support high-bit-depth compiles of x265
     # on 64bit architectures. Main10 plus large resolution plus slow
     # preset plus 32bit address space usually means malloc failure.  You
@@ -419,7 +443,7 @@ if(X64 OR ARM64 OR PPC64)
     # license" so to speak.  If it breaks you get to keep both halves.
     # You will need to disable assembly manually.
     option(HIGH_BIT_DEPTH "Store pixel samples as 16bit values (Main10/Main12)" OFF)
-endif(X64 OR ARM64 OR PPC64)
+endif(X64 OR ARM64 OR PPC64 OR LOONGARCH64)
 if(HIGH_BIT_DEPTH)
     option(MAIN12 "Support Main12 instead of Main10" OFF)
     if(MAIN12)
@@ -523,6 +547,10 @@ if(POWER)
     endif()
 endif()
 
+if(LOONGARCH64)
+    list(APPEND ASM_FLAGS -DHIGH_BIT_DEPTH=0 -DBIT_DEPTH=8 -DX265_NS=${X265_NS})
+endif(LOONGARCH64)
+
 include(Version) # determine X265_VERSION and X265_LATEST_TAG
 include_directories(. common encoder "${PROJECT_BINARY_DIR}")
 
@@ -640,6 +668,19 @@ if((MSVC_IDE OR XCODE OR GCC) AND ENABLE_ASSEMBLY)
                 COMMAND ${NASM_EXECUTABLE} ARGS ${NASM_FLAGS} ${ASM_SRC} -o ${ASM}.${SUFFIX}
                 DEPENDS ${ASM_SRC})
         endforeach()
+    elseif(LOONGARCH64)
+    # compile LOONGARCH arch asm files here
+        enable_language(ASM)
+        foreach(ASM ${LOONGARCH64_ASMS})
+            set(ASM_SRC ${CMAKE_CURRENT_SOURCE_DIR}/common/loongarch64/${ASM})
+            list(APPEND ASM_SRCS ${ASM_SRC})
+            list(APPEND ASM_OBJS ${ASM}.${SUFFIX})
+            add_custom_command(
+                OUTPUT ${ASM}.${SUFFIX}
+                COMMAND ${CMAKE_CXX_COMPILER}
+                ARGS ${LOONGARCH64_ARGS} -c ${ASM_SRC} -o ${ASM}.${SUFFIX}
+                DEPENDS ${ASM_SRC})
+        endforeach()
     endif()
 endif()
 source_group(ASM FILES ${ASM_SRCS})
diff --git a/source/cmake/FindLASX.cmake b/source/cmake/FindLASX.cmake
new file mode 100644
index 000000000..6f635c042
--- /dev/null
+++ b/source/cmake/FindLASX.cmake
@@ -0,0 +1,14 @@
+include(FindPackageHandleStandardArgs)
+
+if(LOONGARCH64)
+
+    set(CHECK_LASX_CODE "
+        int main(int argc, char **argv) {
+            __asm__ volatile (
+               \"xvadd.w $xr0, $xr1, $xr1\"
+            );
+           return 0; }")
+
+    check_cxx_source_compiles("${CHECK_LASX_CODE}" SUPPORTS_LASX)
+
+endif()
diff --git a/source/cmake/FindLSX.cmake b/source/cmake/FindLSX.cmake
new file mode 100644
index 000000000..a7c3e4198
--- /dev/null
+++ b/source/cmake/FindLSX.cmake
@@ -0,0 +1,14 @@
+include(FindPackageHandleStandardArgs)
+
+if(LOONGARCH64)
+
+    set(CHECK_LSX_CODE "
+        int main(int argc, char **argv) {
+            __asm__ volatile (
+               \"vadd.w $vr0, $vr1, $vr1\"
+            );
+            return 0; }")
+
+    check_cxx_source_compiles("${CHECK_LSX_CODE}" SUPPORTS_LSX)
+
+endif()
diff --git a/source/common/CMakeLists.txt b/source/common/CMakeLists.txt
index 184b94127..ec526593b 100644
--- a/source/common/CMakeLists.txt
+++ b/source/common/CMakeLists.txt
@@ -133,7 +133,6 @@ if(POWER)
     endif()
 endif()
 
-
 # set_target_properties can't do list expansion
 string(REPLACE ";" " " VERSION_FLAGS "${VFLAGS}")
 set_source_files_properties(version.cpp PROPERTIES COMPILE_FLAGS ${VERSION_FLAGS})
@@ -144,7 +143,11 @@ if(HAVE_STRTOK_R)
 endif()
 
 if(GCC AND CC_HAS_NO_NARROWING)
-    set_source_files_properties(cpu.cpp PROPERTIES COMPILE_FLAGS -Wno-narrowing)
+    if(ENABLE_ASSEMBLY)
+        set_source_files_properties(cpu.cpp PROPERTIES COMPILE_FLAGS "-DENABLE_ASSEMBLY=1 -Wno-narrowing")
+    else()
+        set_source_files_properties(cpu.cpp PROPERTIES COMPILE_FLAGS -Wno-narrowing)
+    endif()
 endif()
 if(WIN32)
     set(WINXP winxp.h winxp.cpp)
diff --git a/source/common/cpu.cpp b/source/common/cpu.cpp
index 6cf0a5312..fd0e72af8 100644
--- a/source/common/cpu.cpp
+++ b/source/common/cpu.cpp
@@ -117,7 +117,9 @@ const cpu_name_t cpu_names[] =
 #endif
 #elif X265_ARCH_POWER8
     { "Altivec",         X265_CPU_ALTIVEC },
-
+#elif X265_ARCH_LOONGARCH64
+    { "LSX",             X265_CPU_LSX},
+    { "LASX",            X265_CPU_LASX},
 #endif // if X265_ARCH_X86
     { "", 0 },
 };
@@ -414,7 +416,32 @@ uint32_t cpu_detect(bool benableavx512)
 #endif
 }
 
-#else // if X265_ARCH_POWER8
+#elif X265_ARCH_LOONGARCH64
+
+#include <sys/auxv.h>
+
+#define LA_HWCAP_LSX    ( 1U << 4 )
+#define LA_HWCAP_LASX   ( 1U << 5 )
+
+uint32_t cpu_detect( bool benableavx512 )
+{
+    uint32_t flags = 0;
+#if ENABLE_ASSEMBLY
+    uint32_t hwcap = (uint32_t)getauxval( AT_HWCAP );
+#if HAVE_LSX
+    if( hwcap & LA_HWCAP_LSX )
+        flags |= X265_CPU_LSX;
+#endif
+#if HAVE_LASX
+    if( hwcap & LA_HWCAP_LASX )
+        flags |= X265_CPU_LASX;
+#endif
+#endif
+
+    return flags;
+}
+
+#else // if X265_ARCH_LOONGARCH64
 
 uint32_t cpu_detect(bool benableavx512)
 {
diff --git a/source/test/testbench.cpp b/source/test/testbench.cpp
index ddcfd6139..688daa74b 100644
--- a/source/test/testbench.cpp
+++ b/source/test/testbench.cpp
@@ -177,6 +177,8 @@ int main(int argc, char *argv[])
         { "SVE2", X265_CPU_SVE2 },
         { "SVE", X265_CPU_SVE },
         { "FastNeonMRC", X265_CPU_FAST_NEON_MRC },
+        { "LSX", X265_CPU_LSX },
+        { "LASX", X265_CPU_LASX },
         { "", 0 },
     };
 
diff --git a/source/x265.h b/source/x265.h
index 4452526ae..4ebad9a6f 100644
--- a/source/x265.h
+++ b/source/x265.h
@@ -545,6 +545,10 @@ typedef enum
 /* IBM Power8 */
 #define X265_CPU_ALTIVEC         0x0000001
 
+/* LOONGARCH */
+#define X265_CPU_LSX             0x0000001
+#define X265_CPU_LASX            0x0000002
+
 #define X265_MAX_SUBPEL_LEVEL   7
 
 /* Log level */
-- 
2.52.0

